File: app/src/lib/server/multiplayer/joinRoom.ts
import type { Server, Socket } from "socket.io";
import { getRoom } from "../rooms";

export type BaseSocketHandler = [
  socket: Socket,
  consumeRateLimit: (
    cookie_header: string | undefined,
    points: number,
  ) => Promise<true | unknown>,
];

export type withIOHandler = (
  ...args: [...BaseSocketHandler, io: Server]
) => undefined;

type SocketHandler = (...args: BaseSocketHandler) => undefined;

export const joinRoom: SocketHandler = (socket, consumeRateLimit) => {
  socket.on("join_room", async (roomId: unknown) => {
    if (!(await consumeRateLimit(socket.handshake.headers.cookie, 5))) return;
    if (typeof roomId !== "string") return;

    const room = await getRoom(roomId);

    if (room) {
      await socket.join(`roomId/${roomId}`);
      socket.emit("joined_room", roomId);
      return;
    }

    socket.emit("error", "Room not found");
  });
};

File: backend/src/handlers/join_room.rs
use socketioxide::extract::{Data, SocketRef, State};
use tracing::info;

use crate::{redis_client::RedisClient, rooms::room_exists};

pub fn join_room(socket: &SocketRef) {
    socket.on(
        "join_room",
        |socket: SocketRef, Data::<String>(room_id), State::<RedisClient>(client)| async move {
            info!("HI!");
            info!("{:?}", room_id);
            let room = room_exists(&client, &room_id).await;

            info!("{:?}", room);
            if room {
                info!("You joined!");
                let _ = socket.join(format!("roomId/{}", room_id));
                let _ = socket.emit("joined_room", room_id);
                return;
            }

            let _ = socket.emit("error", "Room not found");
        },
    );
}
